### DISCUSSION TITLE
Unexpected behaviour of transport_options.visibility_timeout on Redis broker

### DISCUSSION THREAD

**Discussion by mapa17** [celery/celery#7926] [Q&A]
Hello Celery community,
I have long runnings tasks that are emitted by producers and consumed by workers running in different containers that can be terminated and their work interrupted at any time. I want each container to be working only at one task and in case it gets interrupted i want celery to resubmit the task, so another worker can pick up the job.

In order to not look tasks and use celery for task retry, I am testing a small test setup where the producers emitting tasks that cause a worker to sleep for 30 seconds.

I am using redis as a broker and the workers are configured to be executing with concurrency=1

celeryconfig.py
```python
worker_prefetch_multiplier = 1
```

```python
@app.task(name="sleep_with_celery", bind=True, acks_late=True, reject_on_worker_lost=True)
def sleep_with_celery(self, duration: int):
    logger.info(f'Celery Task will sleep for {duration} seconds ...')
    sleep(duration)
    logger.info(f'Done sleeping!')
```

The producer is configured with 

celeryconfig.py
```python
broker_transport_options = {
    'visibility_timeout': 1*1
}
```

I manually trigger a task on the producer that will send out a single task for the workers, and observe the unacked hash map and the queue (i.e. long) that is holding the tasks

```bash
watch -n 1 redis-cli hgetall unacked
watch -n 1 redis-cli lrange long 0 10
```

I want to make sure that celery will be resubmitting tasks that have been started by workes that have been shutdown/terminated before finishing.
I therefore kill either the complete worker task (celery main thread) or only the subprocessed spawn and observe the queues.

I can see that a task is added to unacked as soon as a worker starts working on it, but instead of being resubmitted once visibility_timeout is reached (for test purposes i set this to 1 sec) it remains in the long queue much longer before its than inserted into the long queue.

## Question
So my problem is that I dont know if i set the visiblity_timeout correctly, as it appears that my settings are ignored.
To make it more concrete, I from the documentation dont understand how broker_transport_options are set correctly in a situation with one redis broker, and multiple producers and consumers. Where, who and in which order the broker_settings must be set? I understand this to be a global or broker specific settings. Do all workers have to have the same value for this setting?

Any clarification is apprechated.

**Comment by humitos** [comment]
I have the exact same problem and I found there are more people with similar issues (see https://github.com/celery/celery/discussions/8959).

I [created a repository to reproduce this issue](https://github.com/humitos/celery-redis-visibility-timeout) when setting `visibility_timeout` to 5s and executing longer tasks (10, 30, 60, and 90 seconds). However, it seems that changing this setting doesn't work for some reason.

@mapa17 Were you able to find a solution to this problem? 🙏🏼

**Comment by mapa17** [comment]
Hi @humitos,

wasting so much time on celery I tried to stay away from it as far as I could. Looking at your repo, I think you have a slightly different use case, having some expectations about the duration of the execution of a task.

Anyways, I looked into it (wasting some more hours on celery) and created an PR: https://github.com/humitos/celery-redis-visibility-timeout/pull/1

Maybe this solves your issues.

Using the app configurations

```python
# Make sure that tasks are acknowledged only once they are completed and timeout will trigger a retry
app.conf.task_acks_on_failure_or_timeout = False
app.conf.task_acks_late = True
```

Will ensure that an task that timed out is going to be retried. In case of your example, indefinitely (i think).

**Reply by humitos** [reply]
Thanks for taking the time on helping me with this.

It seems that I didn't communicated my issue clearly on the repository. I updated its code with comments to explain myself in a better way. I also added `app.conf.task_acks_late = True` that I was missing.

In summary, I found that setting `visibility_timeout=5`, `task_acks_late=True` and triggering a task that takes `30s` to complete does not re-schedule the task at 5s --which is what I'm expecting due the low visibility timeout value and the acks late. It seems the `visibility_timeout` value manually defined is not being respected.

More context: in my real project running in production we have `task_acks_late=True`, `visibility_timeout=7200` and we trigger tasks that take 4500 seconds to complete.

**Comment by mapa17** [comment]
Hi,

Its my understanding, that while a worker process is running, and no task `time_limit` is set, no re-scheduling is happening. 

So based on your comments I am surprised that you see a difference in your tasks based on `visibility_timeout`. For me `visibility_timeout` indicates the duration which it is ok for the system (all workers + broker) to not know about a the status of a worker. So before the `visibility_timeout` if a worker crashes (hard) nothing should happen. Only after that, in case a task was scheduled to that worker that was not acknowledged, would be re-scheduled.

PS: I can not remember the details when looking through the celery codebase, but i think that visibility_timeout was defining how frequent the worker is notifying the broker about it executing a specific task. If the broker gets no updates after `visibility_timeout` i think any next worker polling the broker for tasks will see the "abandoned" task and can perform a re-scheduling.
